### ISSUE TITLE
Clarification about conditioning on lvef and other clinical parameters

### ISSUE THREAD

**Issue by benearnthof** [HReynaud/EchoDiffusion#1]
Hi, I'm currently trying to use this codebase to train a diffusion model for 3D CT images/videos and want to condition on additional clinical parameters.  
I've had a look through the codebase and found that the Dataloader specified [here](https://github.com/HReynaud/EchoDiffusion/blob/main/diffusion/dataset.py#L114) simply returns lvef as a normalized tensor with added batch and channel dimensions.  

My question is now: What operations are performed with this tensor to use it as conditioning information. I could not find any changes to the models which confused me. My goal would be to ideally swap out lvef for a tensor containing other clinical data relevant for my usecase, but I can't figure out where the lvef tensor is "mixed" into the model conditioning info.  

I've also had a thorough look through the [customized imagen branch](https://github.com/lucidrains/imagen-pytorch/compare/main...HReynaud:imagen-pytorch:distillation) that is specified in the requirements but all I could find are the changes to the conditioning image that is supplied to the imagen model and sampled from the input video.   

Could you please clarify how exactly the lvef tensor is used as conditioning info or point out the relevant lines in the code? I'm sure I must have missed this! 

Thanks for the great code, it works flawlessly apart from this confusion!

**Comment by HReynaud** [comment]
Hi @benearnthof,

Thank you for your interest in this work !

The LVEF is indeed sent "as-is" to the imagen trainer, as if it was a text embedding. This works because the imagen model from [imagen-pytorch](https://github.com/lucidrains/imagen-pytorch) will pass it through an [MLP](https://github.com/lucidrains/imagen-pytorch/blob/f9fb8beb69eea15454153226f3a0d8323152d66b/imagen_pytorch/imagen_video.py#L1375) to adjust the embedding (LVEF tensor, B x 1 x 1) to the size expected by the attention layers (B x cond_dim). The MLP basically plays the same role as an embedding layer, and there is therefore no need, theoretically at least, to use more sophisticated methods to pass any clinical parameter.

If you have 3 clinical parameters, like age, sex and weight, you can probably normalize these values to the same range ([0;1] or [-1;1]) and concatenate them as a single tensor B x 1 x 3, and let the MLP learn what each parameter does to the images.

Regarding the code, you can see [here](https://github.com/HReynaud/EchoDiffusion/blob/764f7aaa927662fd75767b352663e45af84d4b09/diffusion/configs/1SCM.yaml#L32) that I tell the trainer to expect ```['images', 'text_embeds', 'cond_images']``` which are the noise input, the LVEF and conditional image, as returned by the dataset.

Let me know if you require further clarifications.

**Comment by benearnthof** [comment]
Thank you for the quick reply! I actually noticed this after digging through the trainer config parameters yesterday night and saw that my problem could be fixed by adjusting the dimensionality to B x 1 x 9 (for 9 relevant parameters). I had figured it must be passed in through the text_embeds some way or another but this is actually quite a clever way of doing it. Thanks for the clarification!
